mongo/integration/client_side_encryption_prose_test.go

```diff
@@ -630,6 +631,81 @@ func TestClientSideEncryptionProse(t *testing.T) {
 			})
 		}
 	})
+	mt.RunOpts("bypass mongocryptd spawning", noClientOpts, func(mt *mtest.T) {
+		kmsProviders := map[string]map[string]interface{}{
+			"local": {
+				"key": localMasterKey,
+			},
+		}
+		schemaMap := map[string]interface{}{
+			"db.coll": readJSONFile(mt, "external-schema.json"),
+		}
+
+		mongocryptdBypassSpawnTrue := map[string]interface{}{
+			"mongocryptdBypassSpawn": true,
+			"mongocryptdURI":         "mongodb://localhost:27021/db?serverSelectionTimeoutMS=1000",
+			"mongocryptdSpawnArgs":   []string{"--pidfilepath=bypass-spawning-mongocryptd.pid", "--port=27021"},
+		}
+		mongocryptdBypassSpawnFalse := map[string]interface{}{
+			"mongocryptdBypassSpawn": false,
+			"mongocryptdSpawnArgs":   []string{"--pidfilepath=bypass-spawning-mongocryptd.pid", "--port=27021"},
+		}
+		mongocryptdBypassSpawnNotSet := map[string]interface{}{
+			"mongocryptdSpawnArgs": []string{"--pidfilepath=bypass-spawning-mongocryptd.pid", "--port=27021"},
+		}
+
+		testCases := []struct {
+			name                    string
+			mongocryptdOpts         map[string]interface{}
+			setBypassAutoEncryption bool
+			bypassAutoEncryption    bool
+		}{
+			{"mongocryptdBypassSpawn only", mongocryptdBypassSpawnTrue, false, false},
+			{"bypassAutoEncryption only", mongocryptdBypassSpawnNotSet, true, true},
+			{"mongocryptdBypassSpawn false, bypassAutoEncryption true", mongocryptdBypassSpawnFalse, true, true},
+			{"mongocryptdBypassSpawn true, bypassAutoEncryption false", mongocryptdBypassSpawnTrue, true, false},
+		}
+		for _, tc := range testCases {
+			mt.Run(tc.name, func(mt *mtest.T) {
+				aeo := options.AutoEncryption().
+					SetKmsProviders(kmsProviders).
+					SetKeyVaultNamespace(kvNamespace).
+					SetSchemaMap(schemaMap).
+					SetExtraOptions(tc.mongocryptdOpts)
+				if tc.setBypassAutoEncryption {
+					aeo.SetBypassAutoEncryption(tc.bypassAutoEncryption)
+				}
+				cpt := setup(mt, aeo, nil, nil)
+				defer cpt.teardown(mt)
+
+				_, err := cpt.cseColl.InsertOne(mtest.Background, bson.D{{"unencrypted", "test"}})
+
+				// Check for mongocryptd server selection error if auto encryption was not bypassed.
+				if !(tc.setBypassAutoEncryption && tc.bypassAutoEncryption) {
+					assert.NotNil(mt, err, "expected InsertOne error, got nil")
+					mcryptErr, ok := err.(mongo.MongocryptdError)
+					assert.True(mt, ok, "expected error type %T, got %v of type %T", mongo.MongocryptdError{}, err, err)
+					assert.True(mt, strings.Contains(mcryptErr.Error(), "server selection error"),
+						"expected mongocryptd server selection error, got %v", err)
+					return
+				}
+
+				// If auto encryption is bypassed, the command should succeed. Create a new client to connect to
+				// mongocryptd and verify it is not running.
+				assert.Nil(mt, err, "InsertOne error: %v", err)
+
+				mcryptOpts := options.Client().ApplyURI("mongodb://localhost:27021").
```

Comment:
Shouldn't the assertions from this line until the end should run regardless of whether auto encryption was bypassed?  If so,  the `return` on line 690 prevents that.  If I'm understanding correctly, the `return` should be removed and the assertion on line 695 should be placed in an `else` block.

Reply:
If auto encryption is not byapssed, it's enough to check that the operation errors out with a mongocryptd server selection error. There's no need to create a new client to check if mongocryptd is running.

These assertions are only run if auto encryption is bypassed because in that case, the `InsertOne` will not communicate with mongocryptd at all and we need to create a new client to check the process.

Reply:
I think this test is confusing as written, but it does seem correct, so I'm going to resolve this thread.